### PR TITLE
Add a new optional argument tid in DataStore.add_entries

### PR DESCRIPTION
**Pull request by feipenghe**
This PR fixes https://github.com/asyml/forte/issues/775 and https://github.com/asyml/forte/issues/781

### Description of changes
* Add a new optional argument tid in DataStore.add_entries
* implemented `add_link_raw` and `add_group_raw`
* implemented `_add_entry_raw` to help modularizing `add_annotation_raw`, `add_link_raw` and `add_group_raw`

### Possible influences of this PR.
Future functions that add new types of entries in raw format can use `_add_entry_raw` 

### Test Conducted
* test `add_link_raw` and `add_group_raw`

### CODE REVIEW

File: forte/data/data_store.py [hunk 1]
```diff
@@ -471,10 +485,18 @@ def add_link_raw(
             ``tid`` of the entry and its index in the ``type_name`` list.
 
         """
-        raise NotImplementedError
+        entry = self._new_link(type_name, parent_tid, child_tid, tid)
+        try:
+            self.__elements[type_name].add(entry)
+        except KeyError:
+            self.__elements[type_name] = SortedList(key=lambda s: (s[0], s[1]))
```
Comment (mylibrar) [inline forte/data/data_store.py:492]: The comparison function for `Link` can be different from the one for `Annotation`. Note that `s[0]` and `s[1]` may have different meanings for `Annotation` entry and `Link` entry.

File: forte/data/data_store.py [hunk 2]
```diff
@@ -489,7 +511,15 @@ def add_group_raw(
             ``tid`` of the entry and its index in the (``type_id``)th list.
 
         """
-        raise NotImplementedError
+        entry = self._new_group(type_name, member_type, tid)
+        try:
+            self.__elements[type_name].add(entry)
+        except KeyError:
+            self.__elements[type_name] = SortedList(key=lambda s: (s[0], s[1]))
```
Comment (mylibrar) [inline forte/data/data_store.py:518]: We may also need to update the comparison function for `Group` entry here.

Comment (feipenghe) [reply forte/data/data_store.py:518]: Where I can find the order that I can refer to?

Comment (mylibrar) [reply forte/data/data_store.py:518]: Previously we use type name as default to compare entries (https://github.com/asyml/forte/blob/master/forte/data/ontology/core.py#L394). Note that some subclasses might overwrite this method (e.g., [Annotation](https://github.com/asyml/forte/blob/master/forte/data/ontology/top.py#L121)).

File: forte/data/data_store.py [hunk 3]
```diff
@@ -425,7 +434,52 @@ def _is_annotation(self, type_name: str) -> bool:
...
+        if entry_type == Annotation:
+            sorting_fn = lambda s: (
+                s[constants.BEGIN_INDEX],
+                s[constants.BEGIN_INDEX],
+            )
+        elif entry_type in [Link, Group]:
+            sorting_fn = lambda s: (s[constants.ENTRY_TYPE_INDEX])
+        else:
+            raise ValueError(
+                f"sorting function for {entry_type} is not" " implemented yet."
```
Comment (hunterhector) [inline forte/data/data_store.py:469]: Merge the two string here (i.e. remove additional ")

File: forte/data/data_store.py [hunk 4]
```diff
@@ -425,7 +434,52 @@ def _is_annotation(self, type_name: str) -> bool:
...
+        Returns:
+            ``tid`` of the entry.
+        """
+        if entry_type == Annotation:
+            sorting_fn = lambda s: (
+                s[constants.BEGIN_INDEX],
+                s[constants.BEGIN_INDEX],
+            )
+        elif entry_type in [Link, Group]:
+            sorting_fn = lambda s: (s[constants.ENTRY_TYPE_INDEX])
```
Comment (hunterhector) [inline forte/data/data_store.py:466]: Actually, I don't think we need sorting for `Link` or `Group`:
1. They are sorted based on the insertion order
2. We don't want to maintain sorted list here
3. Each entry (link or group) are adding to their own list, so the sorting here is not needed (they will be all the same for the same element list)

File: forte/data/data_store.py [hunk 5]
```diff
@@ -425,7 +434,52 @@ def _is_annotation(self, type_name: str) -> bool:
...
+        else:
+            raise ValueError(
+                f"sorting function for {entry_type} is not" " implemented yet."
+            )
+        try:
+            self.__elements[type_name].add(entry)
+        except KeyError:
+            self.__elements[type_name] = SortedList(key=sorting_fn)
```
Comment (hunterhector) [inline forte/data/data_store.py:474]: No sortedlist for link or group

File: forte/data/data_store.py [hunk 6]
```diff
@@ -340,15 +347,17 @@ def _new_link(
...
         entry: List[Any]
 
         entry = [parent_tid, child_tid, tid, type_name]
         entry += self._num_attributes_for_type(type_name) * [None]
 
         return entry
 
-    def _new_group(self, type_name: str, member_type: str) -> List:
+    def _new_group(
+        self, type_name: str, member_type: str, tid: Optional[int] = None
```
Comment (hunterhector) [inline forte/data/data_store.py:359]: add tid to the docstring

File: forte/data/data_store.py [hunk 7]
```diff
@@ -324,7 +327,11 @@ def _new_annotation(self, type_name: str, begin: int, end: int) -> List:
         return entry
 
     def _new_link(
-        self, type_name: str, parent_tid: int, child_tid: int
+        self,
+        type_name: str,
+        parent_tid: int,
+        child_tid: int,
+        tid: Optional[int] = None,
```
Comment (hunterhector) [inline forte/data/data_store.py:334]: add these new parameters to the docstring

File: forte/data/data_store.py [hunk 8]
```diff
@@ -471,10 +522,11 @@ def add_link_raw(
             ``tid`` of the entry and its index in the ``type_name`` list.
 
         """
-        raise NotImplementedError
+        entry = self._new_link(type_name, parent_tid, child_tid, tid)
+        return self._add_entry_raw(Link, type_name, entry)
 
     def add_group_raw(
-        self, type_name: str, member_type: str
+        self, type_name: str, member_type: str, tid: Optional[int] = None
```
Comment (hunterhector) [inline forte/data/data_store.py:529]: add tid to the docstring

File: forte/data/data_store.py (hunk 6 above)
Comment (feipenghe) [reply forte/data/data_store.py:362]: Added.

File: forte/data/data_store.py (hunk 7 above)
Comment (feipenghe) [reply forte/data/data_store.py:336]: Added.

File: forte/data/data_store.py (hunk 8 above)
Comment (feipenghe) [reply forte/data/data_store.py:535]: Added.

File: forte/data/data_store.py [hunk 9]
```diff
@@ -425,7 +438,52 @@ def _is_annotation(self, type_name: str) -> bool:
...
+        based on corresponding type name
+         and sort them based on entry type.
+
+        Args:
+            entry_type: entry's type which decides the sorting of entry.
+            type_name: The name of type in `self.__elements`.
+            entry: raw entry data in the list format.
+
+        Raises:
+            ValueError: raised when the entry type's sorting function is not
```
Comment (hunterhector) [inline forte/data/data_store.py:458]: No `ValueError` is raised in this function, I only see `KeyError`

File: forte/data/data_store.py [hunk 10]
```diff
@@ -452,6 +466,48 @@ def _is_annotation(self, type_name: str) -> bool:
...
+                self.__elements[type_name] = SortedList(key=sorting_fn)
+                self.__elements[type_name].add(entry)
+        elif entry_type in [Link, Group]:
+            try:
+                self.__elements[type_name].append(entry)
+            except KeyError:
+                self.__elements[type_name] = []
+                self.__elements[type_name].append(entry)
+        tid = entry[constants.TID_INDEX]
+        self.__tid_ref_dict[tid] = entry
```
Comment (mylibrar) [inline forte/data/data_store.py:508]: `self.__tid_ref_dict` only stores annotation-like entries. We will need `self.__tid_idx_dict` to keep track of non-annotation-like entries. Refer to https://github.com/asyml/forte/blob/master/forte/data/data_store.py#L198

File: forte/data/data_store.py [hunk 11]
```diff
@@ -452,6 +466,51 @@ def _is_annotation(self, type_name: str) -> bool:
...
+            try:
+                self.__elements[type_name].append(entry)
+            except KeyError:
+                self.__elements[type_name] = []
+                self.__elements[type_name].append(entry)
+        tid = entry[constants.TID_INDEX]
+        if self._is_annotation(type_name):
+            self.__tid_ref_dict[tid] = entry
+        else:
+            self.__tid_idx_dict[tid] = entry
```
Comment (mylibrar) [inline forte/data/data_store.py:511]: `self.__tid_idx_dict` does not directly store the entry data. It only maintains the type name and the corresponding index: https://github.com/asyml/forte/blob/master/forte/data/data_store.py#L205

File: tests/forte/data/data_store_test.py [hunk 12]
```diff
@@ -516,6 +517,26 @@ def test_add_annotation_raw(self):
...
         self.assertEqual(len(self.data_store._DataStore__tid_ref_dict), 8)
 
+    def test_add_link_raw(self):
+        self.data_store.add_link_raw(
+            "forte.data.ontology.top.Link", 9999, 1234567
+        )
+        num_link = len(
+            self.data_store._DataStore__elements["forte.data.ontology.top.Link"]
+        )
+        self.assertEqual(num_link, 2)
```
Comment (hunterhector) [inline tests/forte/data/data_store_test.py:527]: should probably test whether the `tid` is assigned correctly?

File: forte/data/data_store.py [hunk 13]
```diff
@@ -384,12 +396,13 @@ def _new_group(self, type_name: str, member_type: str) -> List:
         Args:
             type_name: The fully qualified type name of the new entry.
             member_type: Fully qualified name of its members.
+            tid: ``tid`` of the ``Group`` entry.
```
Comment (hunterhector) [inline forte/data/data_store.py:399]: This, and other occurrences of `tid`, should expalin that `tid` can be optional and will be auto-assigned.

File: tests/forte/data/data_store_test.py (hunk 12 above)
Comment (feipenghe) [reply tests/forte/data/data_store_test.py:568]: added the test.

**Review by mylibrar — APPROVED** [review]

**Pull request merged**